csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/roles/Roles.kt

```diff
@@ -0,0 +1,113 @@
+import io.github.csaf.sbom.validation.requirements.Requirement17
+import io.github.csaf.sbom.validation.requirements.Requirement18
+import io.github.csaf.sbom.validation.requirements.Requirement19
+import io.github.csaf.sbom.validation.requirements.Requirement20
+import io.github.csaf.sbom.validation.requirements.Requirement21
+import io.github.csaf.sbom.validation.requirements.Requirement22
+import io.github.csaf.sbom.validation.requirements.Requirement23
+import io.github.csaf.sbom.validation.requirements.Requirement5
+import io.github.csaf.sbom.validation.requirements.Requirement6
+import io.github.csaf.sbom.validation.requirements.Requirement7
+import io.github.csaf.sbom.validation.requirements.Requirement8
+import io.github.csaf.sbom.validation.requirements.Requirement9
+import io.github.csaf.sbom.validation.requirements.TlpWhiteAccessible
+import io.github.csaf.sbom.validation.requirements.UsageOfTls
+import io.github.csaf.sbom.validation.requirements.ValidCSAFDocument
+import io.github.csaf.sbom.validation.requirements.ValidFilename
+import io.github.csaf.sbom.validation.requirements.YearInFolder
+
+/**
+ * The "CSAF publisher" role. See
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#721-role-csaf-publisher.
+ */
+object CSAFPublisherRole : Role {
+
+    override val roleRequirements = none()
+
+    override val documentRequirements =
+        allOf(ValidCSAFDocument, ValidFilename, UsageOfTls, TlpWhiteAccessible)
+}
+
+/**
+ * The "CSAF provider" role. See
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#722-role-csaf-provider.
+ */
+object CSAFProviderRole : Role {
+    override val roleRequirements =
+        CSAFPublisherRole.roleRequirements +
+            allOf(Requirement6, Requirement7) +
+            oneOf(Requirement8, Requirement9, Requirement10) +
+            (allOf(YearInFolder, Requirement12, Requirement13, Requirement14) or
+                allOf(Requirement15, Requirement16, Requirement17))
+
+    override val documentRequirements = CSAFPublisherRole.documentRequirements + Requirement5
+}
+
+/**
+ * The "CSAF trusted provider role".
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#723-role-csaf-trusted-provider.
+ */
+object CSAFTrustedProviderRole : Role {
+    override val roleRequirements = CSAFProviderRole.roleRequirements + Requirement20
+
+    override val documentRequirements =
+        CSAFProviderRole.documentRequirements + allOf(Requirement18, Requirement19)
+}
+
+/**
+ * The "CSAF lister role". See
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#724-role-csaf-lister.
+ */
+object CSAFListerRole : Role {
+    override val roleRequirements: Requirement = allOf(Requirement6, Requirement21, Requirement22)
+
+    override val documentRequirements = none()
+}
+
+/**
+ * The "CSAF aggregator role". See
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#725-role-csaf-aggregator.
+ */
+object CSAFAggregatorRole : Role {
```

Comment:
These requirements are currently missing (or not obviously covered):

> * uses the value aggregator for /aggregator/category.
> * lists a mirror for at least two disjoint issuing parties pointing to a domain under its own control.
> * links the public part of the OpenPGP key used to sign CSAF documents for each mirrored issuing party in the corresponding provider-metadata.json.
> * provides for each CSAF document that is mirrored a signature (requirement 19) and a hash (requirement 18). Both SHALL be listed in the ROLIE feed. If the issuing party provides those files for a CSAF document, they SHOULD be copied as well. If the issuing party does not provide those files, they SHALL be created by the CSAF aggregator. Such a signature does not imply any liability of CSAF aggregator for the content of the corresponding CSAF document. It just confirms that the CSAF document provided has not been modified after being downloaded from the issuing party. A CSAF aggregator MAY add additional signatures and hashes for a CSAF document.

Reply:
-> #45